p2panda-auth/src/group/resolver.rs

```diff
@@ -30,70 +42,703 @@ where
     GS: GroupStore<ID, Group = GroupState<ID, OP, C, Self, ORD, GS>> + Debug,
 {
     type State = GroupState<ID, OP, C, Self, ORD, GS>;
-    type Error = GroupResolverError;
+    type Error = GroupResolverError<ID, OP>;
 
-    fn rebuild_required(y: &Self::State, operation: &ORD::Message) -> bool {
+    fn rebuild_required(
+        y: &GroupState<ID, OP, C, Self, ORD, GS>,
+        operation: &ORD::Message,
+    ) -> Result<bool, GroupResolverError<ID, OP>> {
         let control_message = operation.payload();
+        let group_id = control_message.group_id();
+        let _actor = operation.sender();
 
         // Sanity check.
-        if control_message.group_id() != y.group_id {
-            panic!();
+        if y.group_id != group_id {
+            // The operation is not intended for this group.
+            return Err(GroupResolverError::IncorrectGroupId(group_id, y.group_id));
         }
 
-        // Get all current tip operations.
-        //
-        // TODO: should be checking against transitive heads here.
-        let heads = y.heads();
+        let is_concurrent = !get_concurrent_operations(&y.graph, operation.id()).is_empty();
 
-        // Detect concurrent operations by comparing the current heads with the new operations'
-        // dependencies.
-        let is_concurrent = heads != HashSet::from_iter(operation.dependencies().clone());
+        // @TODO: Think we need to check here if there is a concurrent operation which means
```

Comment:
As I say in the comment, I believe we also need to check here for the reverse case, where the concurrent remove/demote already exists, and we need to react to it because of concurrent operations arriving afterwards.